File: platforms/android/app/src/main/java/space/vpnboss/ApiClient.kt
package space.vpnboss

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLDecoder
import java.util.Locale
import android.util.Base64

data class AuthInit(val appCode: String, val authUrl: String)
data class AuthCheck(val status: String, val subscriptionUrl: String?)
data class ProfileState(
    val email: String,
    val needsCompletion: Boolean,
    val hasSubscription: Boolean,
    val trialUsed: Boolean,
)
data class RouteItem(
    val id: Long,
    val flag: String,
    val countryCode: String,
    val name: String,
    val location: String,
    val detail: String,
    val config: String,
)

class ApiClient(private val baseUrl: String = "https://vpnboss.space") {
    var token: String = ""

    fun appAuthInit(): AuthInit {
        val payload = request("POST", "/api/app-auth/init", JSONObject())
        return AuthInit(payload.getString("appCode"), payload.getString("authUrl"))
    }

    fun appAuthCheck(appCode: String): AuthCheck {
        val payload = request("GET", "/api/app-auth/check/$appCode", null, authenticated = false)
        if (payload.optString("status") == "confirmed") token = payload.optString("token")
        return AuthCheck(
            status = payload.optString("status", "pending"),
            subscriptionUrl = payload.optJSONObject("connect")?.optString("subUrl")?.takeIf(String::isNotBlank),
        )
    }

    fun profile(): ProfileState {
        val payload = request("GET", "/api/auth/me", null)
        return ProfileState(
            email = payload.optString("email"),
            needsCompletion = payload.optBoolean("needsProfileCompletion", false),
            hasSubscription = payload.optJSONObject("sub") != null,
            trialUsed = payload.optBoolean("trialUsed", false),
        )
    }

    fun connectionUrl(): String {
        return request("GET", "/api/connect", null).getString("subUrl")
    }

    fun loadRoutes(subscriptionUrl: String): List<RouteItem> {
        require(subscriptionUrl.startsWith("https://")) { "Ссылка подписки должна использовать HTTPS" }
        return loadSubscription(subscriptionUrl).also {
            require(it.isNotEmpty()) { "Подписка пока не содержит доступных серверов" }
        }
    }

    private fun loadSubscription(rawUrl: String): List<RouteItem> {
        val url = if (rawUrl.startsWith("https://")) rawUrl else "https://sekretnik1.vps.webdock.cloud/sub/${rawUrl.removePrefix("/sub/").removePrefix("sub/").trim('/')}"
        val source = requestText(url)
        val decoded = decodeSubscription(source)
        return decoded.lineSequence()
            .map(String::trim)
            .filter { it.startsWith("vless://", ignoreCase = true) }
            .mapIndexedNotNull { index, link -> routeFromLink(index, link) }
            .toList()
    }

    private fun decodeSubscription(source: String): String {
        val trimmed = source.trim()
        if (trimmed.contains("vless://", ignoreCase = true)) return trimmed
        val normalized = trimmed.replace('-', '+').replace('_', '/').filterNot(Char::isWhitespace)
        val padded = normalized + "=".repeat((4 - normalized.length % 4) % 4)
        return runCatching { String(Base64.decode(padded, Base64.DEFAULT), Charsets.UTF_8) }.getOrDefault(trimmed)
    }

    private fun routeFromLink(index: Int, link: String): RouteItem? = runCatching {
        val uri = android.net.Uri.parse(link)
        val fragment = URLDecoder.decode(uri.fragment ?: uri.host ?: "VPNBOSS", "UTF-8")
        val serviceText = fragment.lowercase()
        if (
            listOf(
                "лимит устройств",
                "удалите одно устройство",
                "device limit",
                "remove one device",
            ).any(serviceText::contains)
        ) return null
        val flag = extractFlag(fragment).ifBlank { flagFor(fragment, uri.host.orEmpty()) }
        val countryCode = countryCodeFromFlag(flag).ifBlank { countryCodeFor(fragment, uri.host.orEmpty()) }
        val name = cleanName(fragment, flag)
            .replace(Regex("(?i)\\s*\\|\\s*iPhone\\s*"), "")
            .replace(Regex("\\s+"), " ")
            .trim()
        RouteItem(
            id = index.toLong(),
            flag = flag,
            countryCode = countryCode,
            name = name,
            location = name,
            detail = "${uri.getQueryParameter("security")?.uppercase() ?: "REALITY"} · ${uri.getQueryParameter("type")?.uppercase() ?: "TCP"}",
            config = link,
        )
    }.getOrNull()

    private fun extractFlag(value: String): String {
        val chars = value.codePoints().toArray()
        for (index in 0 until chars.size - 1) {
            if (chars[index] in 0x1F1E6..0x1F1FF && chars[index + 1] in 0x1F1E6..0x1F1FF) {
                return String(chars, index, 2)
            }
        }
        return ""
    }

    private fun cleanName(value: String, flag: String): String = value.replace(flag, "").trim().ifBlank { "VPNBOSS" }

    private fun countryCodeFromFlag(flag: String): String {
        val points = flag.codePoints().toArray()
        if (points.size < 2 || points[0] !in 0x1F1E6..0x1F1FF || points[1] !in 0x1F1E6..0x1F1FF) return ""
        return "${('a'.code + points[0] - 0x1F1E6).toChar()}${('a'.code + points[1] - 0x1F1E6).toChar()}"
    }

    private fun countryCodeFor(vararg values: String): String {
        val value = values.joinToString(" ").lowercase()
        val fixed = when {
            listOf("europe", "europa", "европа", "евросоюз", "european union").any(value::contains) -> "eu"
            listOf("denmark", "дания", "copenhagen", "копенгаген").any(value::contains) -> "dk"
            listOf("germany", "германия", "frankfurt", "берлин").any(value::contains) -> "de"
            listOf("spain", "испания", "madrid", "мадрид").any(value::contains) -> "es"
            listOf("netherlands", "нидерланды", "amsterdam", "амстердам").any(value::contains) -> "nl"
            listOf("finland", "финляндия", "helsinki", "хельсинки").any(value::contains) -> "fi"
            listOf("france", "франция", "paris", "париж").any(value::contains) -> "fr"
            listOf("united kingdom", "great britain", "london", "великобритания", "лондон").any(value::contains) -> "gb"
            listOf("united states", "usa", "new york", "сша").any(value::contains) -> "us"
            listOf("russia", "россия", "moscow", "москва").any(value::contains) -> "ru"
            else -> ""
        }
        if (fixed.isNotEmpty()) return fixed

        val languages = listOf(Locale.ENGLISH, Locale("ru"), Locale("es"))
        for (code in Locale.getISOCountries()) {
            val lowerCode = code.lowercase()
            if (Regex("(^|[^a-z])${Regex.escape(lowerCode)}([^a-z]|$)").containsMatchIn(value)) return lowerCode
            val region = Locale.Builder().setRegion(code).build()
            if (languages.any { locale ->
                    region.getDisplayCountry(locale).lowercase(locale).takeIf { it.length >= 4 }?.let(value::contains) == true
                }) return lowerCode
        }
        return "xx"
    }

    private fun flagFor(vararg values: String): String {
        val value = values.joinToString(" ").lowercase()
        return when {
            listOf("denmark", "дания", "copenhagen", "копенгаген", " dk").any(value::contains) -> "🇩🇰"
            listOf("germany", "германия", "frankfurt", "берлин", " de").any(value::contains) -> "🇩🇪"
            listOf("spain", "испания", "madrid", "мадрид", " es").any(value::contains) -> "🇪🇸"
            listOf("netherlands", "нидерланды", "amsterdam", "амстердам", " nl").any(value::contains) -> "🇳🇱"
            listOf("finland", "финляндия", "helsinki", "хельсинки", " fi").any(value::contains) -> "🇫🇮"
            listOf("france", "франция", "paris", "париж", " fr").any(value::contains) -> "🇫🇷"
            listOf("united kingdom", "london", "великобритания", "лондон", " uk").any(value::contains) -> "🇬🇧"
            listOf("usa", "united states", "сша", "new york", " us").any(value::contains) -> "🇺🇸"
            listOf("russia", "россия", "moscow", "москва", " ru").any(value::contains) -> "🇷🇺"
            else -> "🌐"
        }
    }

    private fun request(method: String, path: String, body: JSONObject?, authenticated: Boolean = true): JSONObject {
        val connection = (URL("$baseUrl$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Content-Type", "application/json; charset=utf-8")
            if (authenticated && token.isNotBlank()) setRequestProperty("Authorization", "Bearer $token")
            if (body != null) {
                doOutput = true
                outputStream.use { it.write(body.toString().toByteArray(Charsets.UTF_8)) }
            }
        }
        val code = connection.responseCode
        val text = (if (code in 200..299) connection.inputStream else connection.errorStream)
            ?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }.orEmpty()
        if (code !in 200..299) {
            val message = runCatching { JSONObject(text).optString("error") }.getOrNull().orEmpty()
            error(message.ifBlank { "Ошибка сервера ($code)" })
        }
        return if (text.isBlank()) JSONObject() else JSONObject(text)
    }

    private fun requestText(url: String): String {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "text/plain, */*")
        }
        val code = connection.responseCode
        val text = (if (code in 200..299) connection.inputStream else connection.errorStream)
            ?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }.orEmpty()
        if (code !in 200..299) error("Подписка недоступна ($code)")
        return text
    }
}

File: platforms/ios/VPNBOSS/APIClient.swift
import Foundation

struct AppAuthInit: Decodable { let appCode: String; let authUrl: String }
struct AppAuthCheck: Decodable { let status: String; let token: String? }
struct UserProfile: Decodable {
    let email: String?
    let hasPassword: Bool?
    let hasEmail: Bool?
    let needsProfileCompletion: Bool?
    let trialUsed: Bool?
    let sub: ActiveSubscription?
}
struct ActiveSubscription: Decodable { let id: Int?; let expiresAt: String? }
struct TrialActivation: Decodable { let ok: Bool; let trial: Bool?; let alreadyActive: Bool?; let expiresAt: String? }
struct ProfileCompletion: Decodable { let ok: Bool }
private struct ConfigBundle: Decodable { let connect: ConnectInfo?; let configs: [ServerConfig]? }
private struct ConnectInfo: Decodable { let subUrl: String? }
private struct ServerConfig: Decodable { let id: Int?; let vlessKey: String?; let name: String?; let server: ServerInfo? }
private struct ServerInfo: Decodable { let name: String?; let location: String? }

struct RouteConfig: Codable, Identifiable, Equatable {
    let id: String
    let flag: String
    let name: String
    let location: String
    let detail: String
    let config: String

    static let placeholder = RouteConfig(id: "placeholder", flag: "🌐", name: "Серверы загружаются", location: "", detail: "Ожидание подписки", config: "")
}

final class APIClient {
    private let base = URL(string: "https://vpnboss.space")!
    var token = ""

    func appAuthInit() async throws -> AppAuthInit { try await request("POST", "/api/app-auth/init", body: [String: String](), authenticated: false) }

    func appAuthCheck(_ code: String) async throws -> AppAuthCheck {
        let result: AppAuthCheck = try await request("GET", "/api/app-auth/check/\(code)", authenticated: false)
        if let token = result.token, !token.isEmpty { self.token = token }
        return result
    }

    func routes() async throws -> [RouteConfig] {
        let bundle: ConfigBundle = try await request("GET", "/api/connect/configs")
        if let raw = bundle.connect?.subUrl, !raw.isEmpty, let routes = try? await subscription(raw), !routes.isEmpty { return routes }
        return (bundle.configs ?? []).compactMap { item in
            guard let link = item.vlessKey, link.lowercased().hasPrefix("vless://") else { return nil }
            let rawName = item.server?.name ?? item.name ?? "VPNBOSS"
            let location = item.server?.location ?? rawName
            let flag = flag(for: location + " " + rawName)
            return RouteConfig(id: String(item.id ?? rawName.hashValue), flag: flag, name: clean(rawName, flag: flag), location: location, detail: "VLESS Reality", config: link)
        }
    }

    func profile() async throws -> UserProfile { try await request("GET", "/api/auth/me") }

    func completeProfile(email: String, password: String) async throws -> ProfileCompletion {
        try await request("POST", "/api/auth/complete-profile", body: ["email": email, "password": password])
    }

    func activateTrial() async throws -> TrialActivation {
        try await request("POST", "/api/trial/activate", body: [String: String]())
    }

    private func subscription(_ raw: String) async throws -> [RouteConfig] {
        let value = raw.hasPrefix("https://") ? raw : "https://sekretnik1.vps.webdock.cloud/sub/\(raw.replacingOccurrences(of: "/sub/", with: "").trimmingCharacters(in: CharacterSet(charactersIn: "/")))"
        let (data, response) = try await URLSession.shared.data(from: URL(string: value)!)
        guard (response as? HTTPURLResponse)?.statusCode ?? 500 < 300 else { throw URLError(.badServerResponse) }
        let source = String(decoding: data, as: UTF8.self).trimmingCharacters(in: .whitespacesAndNewlines)
        let decoded: String
        if source.localizedCaseInsensitiveContains("vless://") { decoded = source }
        else {
            var normalized = source.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/").filter { !$0.isWhitespace }
            normalized += String(repeating: "=", count: (4 - normalized.count % 4) % 4)
            decoded = Data(base64Encoded: normalized).map { String(decoding: $0, as: UTF8.self) } ?? source
        }
        return decoded.split(whereSeparator: \.isNewline).enumerated().compactMap { index, line in route(from: String(line), index: index) }
    }

    private func route(from link: String, index: Int) -> RouteConfig? {
        guard link.lowercased().hasPrefix("vless://"), let components = URLComponents(string: link) else { return nil }
        let fragment = components.fragment?.removingPercentEncoding ?? "VPNBOSS"
        let routeFlag = extractFlag(fragment) ?? flag(for: fragment + " " + (components.host ?? ""))
        var name = clean(fragment, flag: routeFlag).replacingOccurrences(of: "| iPhone", with: "", options: .caseInsensitive).trimmingCharacters(in: .whitespaces)
        if name.isEmpty { name = "VPNBOSS" }
        let query = Dictionary(uniqueKeysWithValues: (components.queryItems ?? []).map { ($0.name, $0.value ?? "") })
        let detail = "\((query["security"] ?? "reality").uppercased()) · \((query["type"] ?? "tcp").uppercased())"
        return RouteConfig(id: "sub-\(index)-\(link.hashValue)", flag: routeFlag, name: name, location: name, detail: detail, config: link)
    }

    private func extractFlag(_ text: String) -> String? {
        let scalars = Array(text.unicodeScalars)
        for i in 0..<(max(0, scalars.count - 1)) where (0x1F1E6...0x1F1FF).contains(Int(scalars[i].value)) && (0x1F1E6...0x1F1FF).contains(Int(scalars[i + 1].value)) {
            return String(String.UnicodeScalarView([scalars[i], scalars[i + 1]]))
        }
        return nil
    }

    private func clean(_ text: String, flag: String) -> String { text.replacingOccurrences(of: flag, with: "").trimmingCharacters(in: .whitespacesAndNewlines) }
    private func flag(for text: String) -> String {
        let value = text.lowercased()
        if ["denmark", "дания", "copenhagen", "копенгаген", " dk"].contains(where: value.contains) { return "🇩🇰" }
        if ["germany", "германия", "frankfurt", "берлин", " de"].contains(where: value.contains) { return "🇩🇪" }
        if ["spain", "испания", "madrid", "мадрид", " es"].contains(where: value.contains) { return "🇪🇸" }
        if ["russia", "россия", "moscow", "москва", " ru"].contains(where: value.contains) { return "🇷🇺" }
        return "🌐"
    }

    private func request<T: Decodable>(_ method: String, _ path: String, body: [String: String]? = nil, authenticated: Bool = true) async throws -> T {
        var request = URLRequest(url: URL(string: path, relativeTo: base)!.absoluteURL)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if authenticated && !token.isEmpty { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body { request.httpBody = try JSONEncoder().encode(body) }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { throw URLError(.badServerResponse) }
        return try JSONDecoder().decode(T.self, from: data)
    }
}
